### PR TITLE
add paddle functions for Matrix ContextProjection APIs

### PR DESCRIPTION
**Pull request by tianbingsz**
added paddle functions for Matrix ContextProjection (https://github.com/PaddlePaddle/Paddle/issues/977)
contextProjectionForward
contextProjectionBackwardData 
contextProjectionBackwardWeight 
contextProjectionBackward

### CODE REVIEW

File: paddle/function/context_projection_op_test.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  auto cpu_weight =
+      is_padding ? std::make_shared<CpuMatrix>(pad, input_dim) : nullptr;
+  auto gpu_weight =
+      is_padding ? std::make_shared<GpuMatrix>(pad, input_dim) : nullptr;
+  if (is_padding) {
+    cpu_weight->randomizeUniform();
+    gpu_weight->copyFrom(*cpu_weight);
+  }
+  IVectorPtr cpu_seq;
+  generateSequenceStartPositions(batch_size, cpu_seq);
```
Comment (hedaoyuan) [inline paddle/function/context_projection_op_test.cpp]: `generateSequenceStartPositions`是一个用于paddle test的通用接口，还是把它单独写一个文件，放到utils编成一个单独的lib，add_library(paddle_test  ....)。这样`add_unittest(context_projection_op_test`里面就不需要显示的增加`../gserver/tests/TestUtil.cpp`。

**Comment by hedaoyuan** [comment]
还需要修改ContextProjection.cpp的实现，这个比较重要，可以用来检查Function的设计是否合理。最终的想法是每个Layer包含一个forward_和一个backward_，Layer::forward/backward只是遍历调用forward_和backward_中的Function。

**Comment by hedaoyuan** [comment]
后续可以一起看一下`Projection`是否可以去掉了，Projection是在MixedLayer里面用的，MixedLayer会将对InputLayer的计算转换成Projection，然后forward/backward时遍历projections_。应该是可以直接将针对InputLayer的计算转换成Function，这样应该就可以去掉Projection了。

File: paddle/gserver/layers/ContextProjection.cpp [hunk 2]
```diff
@@ -78,25 +105,35 @@ LayerStatePtr ContextProjection::getState() {
...
-      config_.context_start(),
-      beginPad_,
-      state_ ? true : isPadding);
+  bool is_padding = config_.trainable_padding();
+  /// first use state_, otherwise use weight_(padding false === w nullptr)
+  auto w_ptr =
+      state_ ? state_.get() : is_padding ? weight_->getW().get() : nullptr;
+  auto start_pos = in_->sequenceStartPositions;
+  /// if use state_ as weight_, w_ptr already has mem, so padding true
+  forward_[0]->init(FuncConfig()
```
Comment (tianbingsz) [inline paddle/gserver/layers/ContextProjection.cpp]: done

File: paddle/function/context_projection_op_test.cpp (hunk 1 above)
Comment (tianbingsz) [inline paddle/function/context_projection_op_test.cpp]: will do it in a separate pull request.

Comment (tianbingsz) [inline paddle/function/context_projection_op_test.cpp]: done, please refer to https://github.com/PaddlePaddle/Paddle/pull/1058

File: paddle/function/ContextProjectionOp.cpp [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+public:
+  void init(const FuncConfig& config) override {
+    context_length_ = config.get<size_t>("context_length");
+    context_start_ = config.get<int>("context_start");
+    begin_pad_ = config.get<size_t>("begin_pad");
+  }
+
+  void calc(const Arguments& inputs,
+            const Arguments& outputs,
+            const Arguments& inouts) override {
```
Comment (emailweixu) [inline paddle/function/ContextProjectionOp.cpp]: Need clear comment about what the function does.
inputs are readonly. outputs are writeonly. inouts are for both read and write. For this function, we should use inouts instead of outputs. Because the result is added to output value.

File: paddle/function/ContextProjectionOp.cpp [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+    }
+  }
+}
+
+/**
+ * \param inputs[0] input grad.
+ * \param inputs[1] weight grad.
+ * \param inputs[2] input sequence.
+ * \param outputs[0] output value.
+ */
```
Comment (emailweixu) [inline paddle/function/ContextProjectionOp.cpp]: comment what the function does

File: paddle/function/ContextProjectionOp.cpp [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    context_length_ = config.get<size_t>("context_length");
+    context_start_ = config.get<int>("context_start");
+    begin_pad_ = config.get<size_t>("begin_pad");
+    is_padding_ = config.get<bool>("is_padding");
+    total_pad_ = config.get<size_t>("total_pad");
+  }
+
+  void calc(const Arguments& inputs,
+            const Arguments& outputs,
+            const Arguments& inouts) override {
```
Comment (emailweixu) [inline paddle/function/ContextProjectionOp.cpp]: should use inouts because gradient are accumulated.

File: paddle/function/ContextProjectionOp.cpp [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+  size_t begin_pad_;
+  bool is_padding_;
+  size_t total_pad_;
+};
+
+/**
+ * \param inputs[0] input grad.
+ * \param inputs[1] input sequence.
+ * \param outputs[0] output grad.
+ */
```
Comment (emailweixu) [inline paddle/function/ContextProjectionOp.cpp]: comment what the function does

File: paddle/function/ContextProjectionOp.cpp [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+class ContextProjectionBackwardDataFunc : public FunctionBase {
+public:
+  void init(const FuncConfig& config) override {
+    context_length_ = config.get<size_t>("context_length");
+    context_start_ = config.get<int>("context_start");
+  }
+
+  void calc(const Arguments& inputs,
+            const Arguments& outputs,
+            const Arguments& inouts) override {
```
Comment (emailweixu) [inline paddle/function/ContextProjectionOp.cpp]: use inouts for accumulate gradient.

File: paddle/function/ContextProjectionOp.cpp [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+  void init(const FuncConfig& config) override {
+    context_length_ = config.get<size_t>("context_length");
+    context_start_ = config.get<int>("context_start");
+    begin_pad_ = config.get<size_t>("begin_pad");
+    total_pad_ = config.get<size_t>("total_pad");
+  }
+
+  void calc(const Arguments& inputs,
+            const Arguments& outputs,
+            const Arguments& inouts) override {
```
Comment (emailweixu) [inline paddle/function/ContextProjectionOp.cpp]: use inouts
